File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<vector<int>> solution(int n) {
    vector<vector<int>> answer(n, vector<int>(n, 0));
    
    int num = 1;
    int cycle = 0;
    while (num <= n*n) {
        for (int i = 0; i < n; i++) {
            if (!answer[cycle][i]) {
                answer[cycle][i] = num;
                num++;
            }
        }

        for (int i = 0; i < n; i++) {
            if (!answer[i][n - 1 - cycle]) {
                answer[i][n - 1 - cycle] = num;
                num++;
            }
        }

        for (int i = n - 1; i >= 0; i--) {
            if (!answer[n - 1 - cycle][i]) {
                answer[n - 1 - cycle][i] = num;
                num++;
            }
        }
        
        for (int i = n - 1; i >= 0; i--) {
            if (!answer[i][cycle]) {
                answer[i][cycle] = num;
                num++;
            }
        }
        
        cycle++;
    }

    return answer;
}

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def first(x, sy, ey, answer):
    global current
    for j in range(sy, ey + 1):
        answer[x][j] = current
        current += 1

def second(y, sx, ex, answer):
    global current
    for i in range(sx, ex + 1):
        answer[i][y] = current
        current += 1
        
def third(x, sy, ey, answer):
    global current
    for j in range(ey, sy - 1, -1):
        answer[x][j] = current
        current += 1
            
def fourth(y, sx, ex, answer):
    global current
    for i in range(ex, sx - 1, -1):
        answer[i][y] = current
        current += 1
        
current = 1

def solution(n):
    answer = [[0] * n for _ in range(n)]
    cycle = 0

    while current <= n * n:
        first(cycle, cycle, n - 1 - cycle, answer)
        second(n - 1 - cycle, cycle + 1, n - 1 - cycle, answer)
        third(n - 1 - cycle, cycle, n - 2 - cycle, answer)
        fourth(cycle, cycle + 1, n - 2 - cycle, answer)
        cycle += 1
    
    for e in answer:
        print(e)
    return answer